### PR TITLE
drop python 3.7 support, drop pypy3.7-3.8, add pypy3.10 (except on windows)

### PR DESCRIPTION
**Pull request by jakkdl**
With 3.7 EOL coming up in 4 days, I thought it time to start writing a PR for dropping it so #2655 can move ahead afterwards.

Only thing I stumbled on so far (until CI tells me otherwise) was `trio/_util:coroutine_or_error` - which seems to be more than an easy search/replace fix.

### CODE REVIEW

File: trio/_core/_tests/tutil.py [hunk 1]
```diff
@@ -86,37 +86,26 @@ def _noop(*args, **kwargs):
...
 
-    @contextmanager
-    def restore_unraisablehook():  # pragma: no cover
-        yield
+@contextmanager
+def disable_threading_excepthook():  # pragma: no cover # not used
+    if sys.version_info >= (3, 10):
+        threading.excepthook, prev = threading.__excepthook__, threading.excepthook
+    else:
+        threading.excepthook, prev = _noop, threading.excepthook
```
Comment (jakkdl) [inline trio/_core/_tests/tutil.py:103]: This is imported in `trio/_core/_tests/test_thread_cache.py` but (surprise surprise!) not used at all. Should I just remove the function completely?

Comment (A5rocks) [reply trio/_core/_tests/tutil.py:103]: I think so. While the `_tests` folder is technically still visible (cause of deprecation) and I got confused by that, this is actually the test folder under `_core` which is totally not visible!

tl;dr no stability promises for this file, feel free to remove. We can always revert if necessary.

Comment (jakkdl) [reply trio/_core/_tests/tutil.py:103]: I was mostly thinking if it "should" be used in tests, or likely to be used in the future, such that we'd want it sticking around.

File: trio/_util.py [hunk 2]
```diff
@@ -139,6 +139,7 @@ def _return_value_looks_like_wrong_library(value):
     # function. So we have to just call it and then check whether the
     # return value is a coroutine object.
     # Note: will not be necessary on python>=3.8, see https://bugs.python.org/issue34890
+    # TODO: python3.7 support is now dropped, so the above can be addressed.
```
Comment (jakkdl) [inline trio/_util.py:142]:
```python
    # We can't check iscoroutinefunction(async_fn), because that will fail
    # for things like functools.partial objects wrapping an async
    # function. So we have to just call it and then check whether the
    # return value is a coroutine object.
    # Note: will not be necessary on python>=3.8, see https://bugs.python.org/issue34890
```

After looking at this for a while, `iscoroutinefunction(async_fn)` breaks 26 tests - and it looks like most, if not all, are due to tests wrapping async methods in sync wrappers - but I think `lowlevel.enable_ki_protection` also breaks it, which seems bad.

Comment (jakkdl) [reply trio/_util.py:142]: Opened #2670 - will leave `coroutine_or_error` alone in this PR.

File: trio/_core/_tests/tutil.py (hunk 1 above)
Comment (A5rocks) [reply trio/_core/_tests/tutil.py]: Oh duh, sorry for missing that.

Looks like it was added for https://github.com/python-trio/trio/commit/fd45d4d8b7694715962b26db948417c02c6cab0b#diff-7ca9928b173566036b4074df65003dc54bd9cedc477b05cd8635ee703da19077R153 but that test was later changed in https://github.com/python-trio/trio/commit/14eb6fe8c2a30a9323ae4dc1901b8b0408c7c1e8 to have some form of it in the main library.

Looks like someone forgot to remove it, so this removal looks good.

**Comment by A5rocks** [comment]
Hi, if I actually am able to release stuff I want to do a patch release with current changes (and without this PR), then soon after a minor release with this PR. This way there's a published version of trio that supports 3.7-3.12 (beta, at least). Sorry it took me a while to figure that out, but that means this PR will have to wait a bit!

**Comment by jakkdl** [comment]
> Hi, if I actually am able to release stuff I want to do a patch release with current changes (and without this PR), then soon after a minor release with this PR. This way there's a published version of trio that supports 3.7-3.12 (beta, at least). Sorry it took me a while to figure that out, but that means this PR will have to wait a bit!

That sounds like a good approach, no worries at all!

**Comment by jakkdl** [comment]
I have no clue what the `notes-to-self` directory is, but ran pyupgrade (and black) on that as well.

**Comment by jakkdl** [comment]
Dropping windows pypy3.10, leaving that for another PR.

**Review by Fuyukai — APPROVED** [review]

**Comment by jakkdl** [comment]
@A5rocks wanna review/merge?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -143,7 +143,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python: ['3.7', '3.8', '3.9', '3.10', 'pypy-3.8-nightly', 'pypy-3.9-nightly']
+        python: ['3.8', '3.9', '3.10', 'pypy-3.9-nightly', 'pypy-3.10-nightly']
```
Comment (A5rocks) [inline .github/workflows/ci.yml:146]: Mind adding at least pypy 3.10 here?

File: README.rst [hunk 4]
```diff
@@ -92,7 +92,7 @@ demonstration of implementing the "Happy Eyeballs" algorithm in an
 older library versus Trio.
 
 **Cool, but will it work on my system?** Probably! As long as you have
-some kind of Python 3.7-or-better (CPython or the latest PyPy3 are
+some kind of Python 3.8-or-better (CPython or currently maintaned versions of PyPy3 are
 both fine), and are using Linux, macOS, Windows, or FreeBSD, then Trio
```
Comment (A5rocks) [inline README.rst:96]: Mind linking to https://doc.pypy.org/en/latest/faq.html#which-python-versions-does-pypy-implement with a `[text](link)` thing here?

Comment (A5rocks) [reply README.rst:96]: The FAQ entry currently has a broken link but that's just a simple issue to fix: I haven't made an issue about it yet though.

File: docs/source/reference-core.rst [hunk 5]
```diff
@@ -974,13 +974,8 @@ work. What we need is something that's *like* a global variable, but
...
 
-To solve this problem, Python 3.7 added a new module to the standard
-library: :mod:`contextvars`. And not only does Trio have built-in
-support for :mod:`contextvars`, but if you're using an earlier version
-of Python, then Trio makes sure that a backported version of
-:mod:`contextvars` is installed. So you can assume :mod:`contextvars`
-is there and works regardless of what version of Python you're using.
-
+To solve this problem, Python has a module in the standard
+library: :mod:`contextvars`.
```
Comment (A5rocks) [inline docs/source/reference-core.rst:978]: I think some whitespace disappeared here?

File: test-requirements.in [hunk 6]
```diff
@@ -3,26 +3,23 @@ pytest >= 5.0         # for faulthandler in core
...
+flake8
 flake8-pyproject
 astor          # code generation
 pip-tools >= 6.13.0
 
 # https://github.com/python-trio/trio/pull/654#issuecomment-420518745
-# typed_ast is deprecated as of 3.8, and straight up doesn't compile on 3.10-dev as of 2021-12-13
-typed_ast; implementation_name == "cpython" and python_version < "3.8"
 mypy-extensions; implementation_name == "cpython"
 typing-extensions < 4.7.0
```
Comment (A5rocks) [inline test-requirements.in:24]: This should be reverted too. (... I forgot to leave a comment on it. Oops.)

File: test-requirements.txt
Comment (A5rocks) [inline test-requirements.txt:1]: Let's make sure to run the autodeps CI thing after this is merged!

File: trio/_core/_tests/tutil.py [hunk 7]
```diff
@@ -43,7 +42,7 @@
     with s:
         try:
             s.bind(("::1", 0))
-        except OSError:
+        except OSError:  # pragma: no cover # since support for 3.7 was removed
```
Comment (A5rocks) [inline trio/_core/_tests/tutil.py:45]: I don't think the `#` between `no cover` and `since support for 3.7 was removed` is necessary, though I actually don't remember this pattern from elsewhere so IDK if trio prefers the more prose-like `# pragma: no cover since support for 3.7 was removed`

**Review by A5rocks — CHANGES_REQUESTED** [review]
Looks mostly good. Please edit `pyproject.toml` to make `black` use `py38` as a minimum version. I don't think it changes anything though.

Sorry about the delay in reviewing! ... I swear I had another thing in mind to ask you to change but whatever, we'll catch it eventually and it won't break anything.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (jakkdl) [reply .github/workflows/ci.yml:146]: In addition to `pypy-3.10-nightly`? Ubuntu tests both `pypy-3.10` and `pypy-3.10-nightly`, windows & mac has always only tested pypy-nightlies (no clue why they do that instead of the stable releases though)

File: trio/_core/_tests/tutil.py (hunk 7 above)
Comment (jakkdl) [reply trio/_core/_tests/tutil.py:45]: `mypy` will error out on action comments with trailing content unless separated by `#`. Idk how codecov handles it, but it seems like a reasonable thing to default to differentiating what has an effect and what's a pure comment.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -18,16 +18,16 @@ jobs:
...
         lsp: ['']
         lsp_extract_file: ['']
         extra_name: ['']
         exclude:
-          - python: 'pypy-3.8-nightly'
-            arch: 'x86'
           - python: 'pypy-3.9-nightly'
             arch: 'x86'
+              #- python: 'pypy-3.10-nightly'
+              #  arch: 'x86'
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply .github/workflows/ci.yml:30]: A quick search indicates that `pypy` straight up doesn't support 32-bit (but I didn't bother trying to find hard documentation asserting that). I was getting errors about missing images (IIRC) in the CI

File: test-requirements.in (hunk 6 above)
Comment (A5rocks) [reply test-requirements.in:24]: Ah, sorry this is pointing towards the typing extensions line!

**Review by A5rocks — CHANGES_REQUESTED** [review]
Just remembered the one more thing: can you make a news fragment?

File: test-requirements.in (hunk 6 above)
Comment (jakkdl) [reply test-requirements.in:24]: Ah, derp. Done!

File: newsfragments/2668.headline.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+Drop support for python3.7, pypy3.7 and pypy3.8.
```
Comment (Fuyukai) [inline newsfragments/2668.headline.rst:1]: should be ``Python 3.7 and PyPy 3.7/3.8`` or similar.

**Review by A5rocks — APPROVED** [review]
This seems fine now.

Except for capitalization in the newsfragment, I suppose.

**Review by A5rocks — APPROVED** [review]
Sorry, I've been thinking about it and IMO this isn't really a "headline" feature; in the past this was a `.removal`.

That plus the capitalization thing are my only concerns, after which feel free to merge!

**Comment by jakkdl** [comment]
> This seems fine now.
> 
> Except for capitalization in the newsfragment, I suppose.

Oh, I missed that capitalization was what Fuyukai actually was requesting be modified. Thanks for clarifying :)

**Comment by jakkdl** [comment]
> Sorry, I've been thinking about it and IMO this isn't really a "headline" feature; in the past this was a `.removal`.
> 
> That plus the capitalization thing are my only concerns, after which feel free to merge!

`removal` wasn't listed in `newsfragments/README.rst` - added it + description it should be used when dropping versions. Feel free to modify that if you want to
